### PR TITLE
added merge strategy that can handle non variants

### PR DESCRIPTION
**Pull request by mhsaul**
Created MergeWithNonVariantsStrategy that can handle merging intervals. Current implementation only supports the same behavior as MoveToCallsStrategy, but functionality will later be extended to add merging logic for non variants.

### CODE REVIEW

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+
+from intervaltree import IntervalTree
+
+from gcp_variant_transforms.beam_io.vcfio import Variant
+from gcp_variant_transforms.libs.variant_merge import move_to_calls_strategy
+from gcp_variant_transforms.libs.variant_merge import variant_merge_strategy
+
+__all__ = ['MergeWithNonVariantsStrategy']
+
+DEFAULT_WINDOW_SIZE = 2500
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py]: Please add a comment about what this value does.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+
+
+class MergeWithNonVariantsStrategy(variant_merge_strategy.VariantMergeStrategy):
+  """A merging strategy that can handle non-variant records.
+  Variants will be merged across files by using reference_name and start
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py]: nit: please add an empty line after the first line.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+  `copy_filter_to_calls` options.
+  Note: if a field is set to be moved from INFO to calls, then it must
+  not already exist in calls (i.e. specified by FORMAT in the VCF
+  header).
+  """
+
+  def __init__(self, info_keys_to_move_to_calls_regex, copy_quality_to_calls,
+               copy_filter_to_calls, window_size=DEFAULT_WINDOW_SIZE):
+    """Initializes the strategy.
+    Args:
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py]: nit: add an empty line here as well.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+      window_size (int): Size of windows that variants will be grouped
+        in based on the start position of the variant
+    """
+    self._move_to_calls_strategy = move_to_calls_strategy.MoveToCallsStrategy(
+        info_keys_to_move_to_calls_regex=info_keys_to_move_to_calls_regex,
+        copy_quality_to_calls=copy_filter_to_calls,
+        copy_filter_to_calls=copy_filter_to_calls)
+    self._window_size = window_size
+
+  def _union(self, list1, list2):
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py]: s/_union/_merge_lists?

File: gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py [hunk 5]
```diff
@@ -60,6 +60,39 @@ def __init__(self, info_keys_to_move_to_calls_regex, copy_quality_to_calls,
     self._copy_quality_to_calls = copy_quality_to_calls
     self._copy_filter_to_calls = copy_filter_to_calls
 
+  def move_data_to_calls(self, variant):
+    """Moves filters, calls, and info items to the variant's calls based on the
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py]: nit: Please ensure the first comment fits in the first line (required by lint).

File: gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py [hunk 6]
```diff
@@ -60,6 +60,39 @@ def __init__(self, info_keys_to_move_to_calls_regex, copy_quality_to_calls,
...
+    strategy's initialization parameters.
+
+    Args:
+      variant (:class: `~vcfio.Variant`): The variant whose filters, quality,
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py]: nit: extra space after :class: ... also, I'd remove the :class: notation from here as we dont use that notation for anything other than beam_io (it's better to be consistent for now until we got back and fix everything)....so just `Variant` should be good.

File: gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py [hunk 7]
```diff
@@ -73,23 +106,12 @@ def get_merged_variants(self, variants):
...
-        additional_call_info[ColumnKeyConstants.QUALITY] = variant.quality
-      for info_key, info_value in variant.info.iteritems():
-        if self._should_move_info_key_to_calls(info_key):
-          additional_call_info[info_key] = info_value.data
-        else:
-          merged_variant.info[info_key] = info_value
-      for call in variant.calls:
-        call.info.update(additional_call_info)
-        merged_variant.calls.append(call)
+      merged_variant.quality = max(merged_variant.quality, variant.quality)
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py]: nice! I didn't realize max works with None types :)

File: gcp_variant_transforms/vcf_to_bq.py [hunk 8]
```diff
@@ -67,6 +76,11 @@ def _get_variant_merge_strategy(known_args):
         known_args.info_keys_to_move_to_calls_regex,
         known_args.copy_quality_to_calls,
         known_args.copy_filter_to_calls)
+  elif known_args.variant_merge_strategy == 'MERGE_WITH_NON_VARIANTS':
```
Comment (arostamianfar) [inline gcp_variant_transforms/vcf_to_bq.py]: Please add this to _validate_args as well.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    return sorted(set(list1 + list2))
+
+  def _merge(self, v1, v2):
+    if v1.alternate_bases == v2.alternate_bases:
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py]: Please check start and end as well (also add a comment that we assume reference_bases match if start and end are the same).

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+                               calls=sorted(v1.calls + v2.calls))
+      self._move_to_calls_strategy.move_data_to_merged(v1, merged_variant)
+      self._move_to_calls_strategy.move_data_to_merged(v2, merged_variant)
+      yield merged_variant
+    else:
+      yield v1
+      yield v2
+
+  def get_merged_variants(self, variants):
+    tree = IntervalTree()
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py]: Please add a TODO for handling non-variant regions. It's not obvious that this is WIP :)

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py (hunk 2 above)
Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py]: Done.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py (hunk 3 above)
Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py]: Done.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py (hunk 4 above)
Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py]: Done.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py (hunk 9 above)
Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py]: Done.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py (hunk 10 above)
Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py]: Done.

File: gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py (hunk 5 above)
Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py]: Done.

File: gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py (hunk 6 above)
Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py]: Done.

File: gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py (hunk 7 above)
Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py]: Yeah it's pretty convenient! I think comparison operators work for most python built in types.

File: gcp_variant_transforms/vcf_to_bq.py (hunk 8 above)
Comment (mhsaul) [reply gcp_variant_transforms/vcf_to_bq.py]: Done.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py (hunk 1 above)
Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy.py]: Done.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy_test.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Tests for merge_with_nonvariants_strategy."""
+
+from __future__ import absolute_import
+
+import unittest
+
+
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy_test.py]: nit: remove extra empty line.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy_test.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+from gcp_variant_transforms.beam_io import vcfio
+from gcp_variant_transforms.libs.bigquery_vcf_schema import ColumnKeyConstants
+from gcp_variant_transforms.libs.variant_merge import (
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy_test.py]: nit: long import statements are actually exempt from line length, so you can keep it as one line.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy_test.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+    strategy = merge_with_non_variants_strategy.MergeWithNonVariantsStrategy(
+        None, None, None, 2)
+
+    variant_1 = vcfio.Variant(reference_name='1', start=3, end=4)
+    variant_2 = vcfio.Variant(reference_name='2', start=4, end=5)
+
+    self.assertEqual(next(strategy.get_merge_keys(variant_1)), '1:2')
+    self.assertEqual(next(strategy.get_merge_keys(variant_2)), '2:4')
+
+  def test_get_nonv_merge_keys(self):
```
Comment (arostamianfar) [inline gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy_test.py]: nit: please use non_variant in the name.

File: gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py (hunk 6 above)
Comment (arostamianfar) [reply gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py]: Was this file not updated?

Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/move_to_calls_strategy.py]: Sorry, looks like I fixed it everywhere but here... Done now. :)

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy_test.py (hunk 11 above)
Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy_test.py]: Done.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy_test.py (hunk 12 above)
Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy_test.py]: Done.

File: gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy_test.py (hunk 13 above)
Comment (mhsaul) [reply gcp_variant_transforms/libs/variant_merge/merge_with_non_variants_strategy_test.py]: Done.

**Pull request merged**